### PR TITLE
Introduce start and end arguments for dimensions

### PR DESCRIPTION
**Pull request by navjotk**
Finally, this is the critical PR that introduces the new devito feature from #247 , letting the user specify the start and end for any dimension. 

This is PR 2/3 on the way to checkpointing.

### CODE REVIEW

**Comment by FabioLuporini** [comment]
I'll review this properly once rebased on master+#342 

However, skimming through it, I can already anticipate this lacks testing. We need many

File: devito/arguments.py [hunk 1]
```diff
@@ -193,14 +189,15 @@ class DimensionArgProvider(ArgumentProvider):
...
     def reset(self):
-        self._value = self._default_value
+        for i in self.rtargs:
+            i.reset()
 
     @property
     def value(self):
-        return self._value
+        child_values = tuple([i.value for i in self.rtargs])
+        return child_values if all(i is not None for i in child_values) else None
```
Comment (FabioLuporini) [inline devito/arguments.py]: In what case can they all be ``None`` ?

File: devito/arguments.py [hunk 2]
```diff
@@ -209,8 +206,23 @@ def dtype(self):
...
     @cached_property
     def rtargs(self):
-        size = ScalarArgument("%s_size" % self.name, self, max)
-        return [size]
+        size = ScalarArgument(self.size_name, self, max)
+        start = ScalarArgument(self.start_name, self, max, 0)
+        end = ScalarArgument(self.end_name, self, max)
+        return [size, start, end]
+
+    def promote(self, value):
```
Comment (FabioLuporini) [inline devito/arguments.py]: docstring

File: devito/arguments.py [hunk 3]
```diff
@@ -209,8 +206,23 @@ def dtype(self):
...
+        if not isinstance(value, tuple):
+            size, start, end = self.rtargs
+            value = (value, start.default_value, value)
```
Comment (FabioLuporini) [inline devito/arguments.py]: if not a tuple, then it's something else, and we're returning it. What is it supposed to be? can we assert it?

File: devito/arguments.py [hunk 4]
```diff
@@ -209,8 +206,23 @@ def dtype(self):
...
+        else:
+            if len(value) == 2:
+                # Assume we've been passed a (start, end) tuple
```
Comment (FabioLuporini) [inline devito/arguments.py]: is "Assume" the right word ? isn't it always the case if we have a 2-tuple ?

File: devito/arguments.py [hunk 5]
```diff
@@ -209,8 +206,23 @@ def dtype(self):
...
+        size = ScalarArgument(self.size_name, self, max)
+        start = ScalarArgument(self.start_name, self, max, 0)
+        end = ScalarArgument(self.end_name, self, max)
+        return [size, start, end]
+
+    def promote(self, value):
+        if not isinstance(value, tuple):
+            size, start, end = self.rtargs
+            value = (value, start.default_value, value)
+        else:
```
Comment (FabioLuporini) [inline devito/arguments.py]: can we just have ``elif``s ? are we handling all cases ?

File: devito/arguments.py [hunk 6]
```diff
@@ -226,10 +238,13 @@ def verify(self, value):
             except AttributeError:
                 return False
 
+        value = self.promote(value)
```
Comment (FabioLuporini) [inline devito/arguments.py]: Probably we need a comment here

File: devito/dimension.py [hunk 7]
```diff
@@ -32,6 +32,30 @@ def symbolic_size(self):
         """The symbolic size of this dimension."""
         return self.rtargs[0].as_symbol
 
+    @property
+    def symbolic_extent(self):
+        """Return the extent of the loop over this dimension.
```
Comment (FabioLuporini) [inline devito/dimension.py]: I wouldn't use the work "loop" here

Comment (FabioLuporini) [reply devito/dimension.py]: BTW, random thought, I think we should start calling these objects ``Iterators``, or at least something better than ``Dimension``. Or we should derive Iterators from Dimensions... don't know. Just a random thought

File: devito/operator.py [hunk 8]
```diff
@@ -231,7 +231,7 @@ def compile(self):
         """
         if self._lib is None:
             # No need to recompile if a shared object has already been loaded.
-            return jit_compile(self.ccode, self._compiler)
+            return jit_compile(self.ccode, configuration['compiler'])
```
Comment (FabioLuporini) [inline devito/operator.py]: This won't work with YASK. It needs to be ``self._compiler``. Perhaps this is post-merge/post-rebase leftover/error

File: devito/operator.py [hunk 9]
```diff
@@ -240,7 +240,7 @@ def cfunction(self):
         """Returns the JIT-compiled C function as a ctypes.FuncPtr object."""
         if self._lib is None:
             basename = self.compile
-            self._lib = load(basename, self._compiler)
+            self._lib = load(basename, configuration['compiler'])
```
Comment (FabioLuporini) [inline devito/operator.py]: same

File: devito/nodes.py [hunk 10]
```diff
@@ -394,6 +394,7 @@ def bounds(self, start=None, finish=None):
         available (either statically known or provided through ``start``/
         ``finish``). ``None`` is used as a placeholder in the returned 2-tuple
         if a limit is unknown."""
+        lower = None
```
Comment (FabioLuporini) [inline devito/nodes.py]: likewise ``upper = None`` ?

File: devito/operator.py [hunk 11]
```diff
@@ -456,7 +456,7 @@ def apply(self, **kwargs):
         self.cfunction(*list(arguments.values()))
 
         # Output summary of performance achieved
-        summary = self.profiler.summary(dim_sizes, self.dtype)
+        summary = self.profiler.summary(arguments, self.dtype)
```
Comment (FabioLuporini) [inline devito/operator.py]: so can we drop ``dim_sizes`` ?

File: devito/profiling.py [hunk 12]
```diff
@@ -137,17 +137,25 @@ def summary(self, dim_sizes, dtype):
...
+                else:
+                    start = arguments[dims[i].start_name]
+                    end = arguments[dims[i].end_name]
+                itershape.append(i.extent(finish=end, start=start))
             iterspace = reduce(operator.mul, itershape)
             flops = float(profile.ops*iterspace)
             gflops = flops/10**9
 
             # Compulsory traffic
             datashape = [i.dim.size if i.dim.is_Fixed
```
Comment (FabioLuporini) [inline devito/profiling.py]: perhaps this is movable inside the ``for i in itspace`` loop above?

File: devito/arguments.py (hunk 1 above)
Comment (navjotk) [reply devito/arguments.py]: Each runtime argument is independent and gets its value independent of the others. When it doesn't have a value, it is `None`, so they can all be `None` if none of them have a value. Is that your question? 
The code above, however, returns a tuple of values if all arguments have a known value, otherwise it returns `None`.

File: devito/arguments.py (hunk 5 above)
Comment (navjotk) [reply devito/arguments.py]: The condition is on whether `value` is a `tuple` or not so technically, yes, we are handling all the cases. What is your concern here?

File: devito/arguments.py (hunk 4 above)
Comment (navjotk) [reply devito/arguments.py]: There are many ways to interpret a 2-tuple in order to promote it to a 3-tuple. I just chose one as convention and stated that assumption here. It is an assumption.

File: devito/arguments.py (hunk 6 above)
Comment (navjotk) [reply devito/arguments.py]: Saying what?

File: devito/dimension.py (hunk 7 above)
Comment (navjotk) [reply devito/dimension.py]: We already derive `Iteration`s from `Dimension`s. How are `Iterator`s different from `Iteration`s?

File: devito/operator.py (hunk 8 above)
Comment (navjotk) [reply devito/operator.py]: Indeed

File: devito/operator.py (hunk 11 above)
Comment (navjotk) [reply devito/operator.py]: I am working towards dropping it, yes. I think this might be controversial with @mloubout though. I believe `arguments`, the other return value has everything we need.

File: devito/nodes.py (hunk 10 above)
Comment (navjotk) [reply devito/nodes.py]: Tests pass without it. They don't pass without this. Something missing in the tests? :)

File: devito/arguments.py (hunk 3 above)
Comment (navjotk) [reply devito/arguments.py]: This method is promoting a scalar to a tuple so everything that is not a tuple is treated like a scalar.

File: devito/arguments.py (hunk 5 above)
Comment (FabioLuporini) [reply devito/arguments.py]: gotcha now

File: devito/arguments.py (hunk 3 above)
Comment (FabioLuporini) [reply devito/arguments.py]: OK. Perhaps make it private with ``_`` ?

File: devito/arguments.py (hunk 1 above)
Comment (FabioLuporini) [reply devito/arguments.py]: I **think** so... unless it's too complicated, could you sketch (pseudocode is fine) a concrete example in which this happens ?

File: devito/arguments.py (hunk 6 above)
Comment (FabioLuporini) [reply devito/arguments.py]: when/why we need promotion

File: devito/nodes.py (hunk 10 above)
Comment (FabioLuporini) [reply devito/nodes.py]: perhaps...

File: devito/arguments.py (hunk 1 above)
Comment (navjotk) [reply devito/arguments.py]: Let's say we are talking of the `time` dimension, it has three `Argument`s - `time_size`, `time_start`, `time_end`. If `time_size.value=100`, `time_start.value=50` and `time_end.value=60`, this method would return `(100, 50, 60)`. If any of them doesn't have a value (and hence is `None`), this method would return `None`.

File: devito/profiling.py (hunk 12 above)
Comment (navjotk) [reply devito/profiling.py]: TBH I don't understand the reason for calculating the datashape and iterspace separately. They seem to resolve to the same number always. Either that or my changes to the code in this section are based on incorrect assumptions.

Comment (FabioLuporini) [reply devito/profiling.py]: the itershape differs in general because of the stencil

Comment (navjotk) [reply devito/profiling.py]: The itershape is the product of the size of all dimensions, and so is the datashape. Correct? Always?

Comment (FabioLuporini) [reply devito/profiling.py]: no it's the product of the extents. Previously ``dim_sizes`` was being used, which in turn was used to call ``extents``

Comment (navjotk) [reply devito/profiling.py]: Which one? Itershape or datashape? What is the other one?

Comment (FabioLuporini) [reply devito/profiling.py]: itershape

File: devito/operator.py (hunk 11 above)
Comment (mloubout) [reply devito/operator.py]: What d you mean drop dim sizes?

File: devito/profiling.py (hunk 12 above)
Comment (navjotk) [reply devito/profiling.py]: And how is datashape different?

Comment (FabioLuporini) [reply devito/profiling.py]: You're touching boundary data through the stencil

Comment (FabioLuporini) [reply devito/profiling.py]: Pls add a comment

File: devito/dimension.py (hunk 7 above)
Comment (FabioLuporini) [reply devito/dimension.py]: It's just that at this point I'm not sure I like the name ``Dimension`` given what these objects can do...but this is more than debatable

**Comment by navjotk** [comment]
Temporarily closing

File: devito/arguments.py (hunk 1 above)
Comment (FabioLuporini) [reply devito/arguments.py]: You wrote:

> Each runtime argument is independent and gets its value independent of the others. When it doesn't have a value, it is None, so they can all be None if none of them have a value. Is that your question?
> The code above, however, returns a tuple of values if all arguments have a known value, otherwise it returns None.
> 

Do you think it'd be useful to add it as a comment in the code ?

File: devito/operator.py [hunk 13]
```diff
@@ -81,7 +81,6 @@ def __init__(self, expressions, **kwargs):
         self.dtype = self._retrieve_dtype(expressions)
         self.input, self.output, self.dimensions = self._retrieve_symbols(expressions)
         stencils = self._retrieve_stencils(expressions)
-
```
Comment (FabioLuporini) [inline devito/operator.py]: please leave a blank line

File: devito/operator.py [hunk 14]
```diff
@@ -151,14 +150,22 @@ def arguments(self, **kwargs):
...
+        runtime_dimensions = [d for d in self.dimensions if not d.is_Fixed]
         for d in runtime_dimensions:
             d.verify(kwargs.pop(d.name, None))
         for i in self.parameters:
             if i.is_ScalarArgument:
                 i.verify(kwargs.pop(i.name, None))
-
-        dim_sizes = OrderedDict([(d.name, d.value) for d in runtime_dimensions])
+        dim_sizes = {}
+        for d in runtime_dimensions:
```
Comment (FabioLuporini) [inline devito/operator.py]: what's going on here?

File: devito/operator.py [hunk 15]
```diff
@@ -151,14 +150,22 @@ def arguments(self, **kwargs):
...
+            if d.value is not None:
+                _, d_start, d_end = d.value
+                # The variable is called size but what it expects is extent
+                # Since this is used for loop blocking
```
Comment (FabioLuporini) [inline devito/operator.py]: I don't like the fact that ``Operator.arguments`` does anything specific to "loop blocking". Can we abstract things a little bit ?

File: devito/profiling.py (hunk 12 above)
Comment (FabioLuporini) [reply devito/profiling.py]: reminder that this still needs to be addressed.

Also, since you're touching the profiling engine, can you verify that, for example, acoustic or tti still produce the same "performance summary" when running from ``master`` and from this ``branch`` ? Can you paste here some data?

File: devito/dimension.py [hunk 16]
```diff
@@ -36,6 +36,30 @@ def symbolic_size(self):
     def size(self):
         return None
 
+    @property
+    def symbolic_extent(self):
+        """Return the extent of the loop over this dimension.
+        Would be the same as size if using default values """
+        _, start, end = self.rtargs
```
Comment (FabioLuporini) [inline devito/dimension.py]: I still think  ``namedtuple`` would be better

File: devito/operator.py (hunk 15 above)
Comment (navjotk) [reply devito/operator.py]: It is not specifically for loop blocking. It is simply calculating the extent of the dimension - which is a general feature.

File: devito/arguments.py [hunk 17]
```diff
@@ -299,7 +339,27 @@ class ObjectArgProvider(ArgumentProvider):
 
     @cached_property
     def rtargs(self):
-        return [PtrArgument(self.name, self)]
+        return (PtrArgument(self.name, self),)
+
+
+class ArgumentEngine(object):
+    def extract_dimension_offsets(self, stencils):
```
Comment (FabioLuporini) [inline devito/arguments.py]: what's going on here (docs) ?

Comment (FabioLuporini) [reply devito/arguments.py]: I might have a vague idea, one or two concrete examples would be of help

File: devito/operator.py [hunk 18]
```diff
@@ -150,15 +157,22 @@ def arguments(self, **kwargs):
...
-            d.verify(kwargs.pop(d.name, None))
+            d.verify(kwargs.pop(d.name, None), self.argument_engine, enforce=True)
         for i in self.parameters:
             if i.is_ScalarArgument:
-                i.verify(kwargs.pop(i.name, None))
-
-        dim_sizes = OrderedDict([(d.name, d.value) for d in runtime_dimensions])
+                i.verify(kwargs.pop(i.name, None), self.argument_engine, enforce=True)
+        dim_sizes = {}
+        for d in runtime_dimensions:
```
Comment (FabioLuporini) [inline devito/operator.py]: perhaps this ``extent`` calculation could be a ``Dimension`` property?

File: devito/operator.py [hunk 19]
```diff
@@ -94,6 +95,12 @@ def __init__(self, expressions, **kwargs):
         # Wrap expressions with Iterations according to dimensions
         nodes = self._schedule_expressions(clusters)
 
+        # Initialise Argument Engine
+        self.argument_engine = ArgumentEngine()
+
+        # Extract dimension offsets
+        self.argument_engine.extract_dimension_offsets([i.stencil for i in clusters])
```
Comment (FabioLuporini) [inline devito/operator.py]: with this ``ArgumentEngine``, are you simply trying to extract the outermost offsets along each dimension? An example; given the expression
```
A[i-3, j+1] - B[i+4, j+5] + C[i-2, j+3]
```
Return:
```
i: (-3, 4)
j: (1, 5)
```

File: devito/operator.py [hunk 20]
```diff
@@ -304,7 +316,8 @@ def _schedule_expressions(self, clusters):
                 needed = entries[index:]
 
                 # Build and insert the required Iterations
-                iters = [Iteration([], j.dim, j.dim.symbolic_size, offsets=j.ofs)
+                iters = [Iteration([], j.dim, j.dim.limits,
+                                   offsets=j.ofs if j.dim.is_Fixed else (0, 0))
```
Comment (FabioLuporini) [inline devito/operator.py]: why?

File: tests/test_operator.py [hunk 21]
```diff
@@ -289,6 +303,63 @@ def test_override_composite_data(self):
         arg_name = src1.name + "_coords"
         assert(np.array_equal(args[arg_name], np.asarray((new_coords,))))
 
+    def test_start_end_2d(self):
```
Comment (FabioLuporini) [inline tests/test_operator.py]: docstrings, for this and all other new tests

File: devito/arguments.py [hunk 22]
```diff
@@ -52,37 +53,38 @@ def as_symbol(self):
...
 
+    def __repr__(self):
+        return self.name
+
 
 class ScalarArgument(Argument):
 
     """ Class representing scalar arguments that a kernel might expect.
         Most commonly used to pass dimension sizes
+        enforce determines whether any reduction will be performed or not.
```
Comment (FabioLuporini) [inline devito/arguments.py]: use doc-friendly notation:
```
:param enforce: Determine ...
```
Also, add an example? Not straightforward to imagine what's the use case here

Comment (FabioLuporini) [reply devito/arguments.py]: Q: should we provide a difference type of ``Argument``, rather than adding a parameter such as ``enforce`` ?

File: devito/arguments.py [hunk 23]
```diff
@@ -52,37 +53,38 @@ def as_symbol(self):
...
 
     is_ScalarArgument = True
 
-    def __init__(self, name, provider, reducer, default_value=None):
+    def __init__(self, name, provider, reducer=lambda old, new: new, default_value=None):
         super(ScalarArgument, self).__init__(name, provider, default_value)
         self.reducer = reducer
 
-    def verify(self, value):
+    def verify(self, value, engine, enforce=False):
```
Comment (FabioLuporini) [inline devito/arguments.py]: unused ``engine``

File: devito/operator.py (hunk 18 above)
Comment (navjotk) [reply devito/operator.py]: Dimensions are global objects shared between different Operators. This is an intrinsic property of an Operator.

File: devito/operator.py (hunk 20 above)
Comment (navjotk) [reply devito/operator.py]: If you're hardcoding the limits, hard-code the offsets as well. Else, the offset will be accounted for in the runtime limit calculation.

Comment (FabioLuporini) [reply devito/operator.py]: no but I mean, why `j.ofs` wouldn't return ``(0, 0)`` if ``j.is_Fixed`` is True?

File: devito/operator.py (hunk 18 above)
Comment (FabioLuporini) [reply devito/operator.py]: I disagree; that'd be a property of a *specific Operator invocation*, not of the Operator itself. And also, so would be ``.limits``. Finally, let's not forget that a Dimension **is** an ArgProvider, so... We can argue whether that should be the case, but that's another story

File: devito/operator.py (hunk 20 above)
Comment (navjotk) [reply devito/operator.py]: Because it would return the offset in that dimension which depends on the stencil radius?

File: devito/operator.py (hunk 18 above)
Comment (navjotk) [reply devito/operator.py]: So if it is a property of the invocation, it belongs right here, no? 
I don't understand what that ArgProvider thing is about but it seems off topic

File: devito/operator.py (hunk 19 above)
Comment (navjotk) [reply devito/operator.py]: Is there a question?

File: devito/arguments.py (hunk 23 above)
Comment (navjotk) [reply devito/arguments.py]: For consistency, all verify methods accept engine.

File: devito/arguments.py (hunk 22 above)
Comment (navjotk) [reply devito/arguments.py]: I don't think your suggestion has anything to do with what is happening here. The argument decides whether the reducer will be called or not. It is still a scalar, but when I am auto-deriving, I want to use a reducer to get the value that everyone agrees on, while if I am using a user-provided value, I want to skip all the reduction operations and just use the provided value directly.

File: devito/operator.py (hunk 19 above)
Comment (FabioLuporini) [reply devito/operator.py]: yes a confirmation that this what you're trying to do

Comment (navjotk) [reply devito/operator.py]: Yes but it doesn't return those values, it caches them on the engine object.

File: devito/operator.py (hunk 20 above)
Comment (FabioLuporini) [reply devito/operator.py]: anyway, we can drop this because of #373

File: devito/operator.py (hunk 19 above)
Comment (FabioLuporini) [reply devito/operator.py]: OK, then I think there should be ways of doing this without this engine, which is only over-complicating things

a few lines above you have the ``stencils`` variable, which you can inspect to get the max/min offsets along any Operator dimension (you might even add helper methods to ``Stencil`` directly?).  What happens if you try using this?

File: devito/operator.py (hunk 18 above)
Comment (FabioLuporini) [reply devito/operator.py]: Being an ``ArgProvider``, it seemed kinda natural to have an ``extent`` method.

I'm just concerned that the ``.arguments`` method is becoming more and more complicated, so it would be useful (for many purposes) to "offload" some bit of functionalities to other modules, if/where possible. Mine was just a suggestion in this direction (perhaps a conceptually wrong one, but I'm not that sure).

**Comment by navjotk** [comment]
The biggest problem in this branch currently is that `test_dse.py/test_tti_rewrite_speculative` fails because there are some hidden assumptions about all loops running from 0 to `dimension_size` in the speculative rewrite passes of the DSE that I haven't been able to find after a week of looking. Can @FabioLuporini please look at making this test pass? More specifically, this problem is most probably localised within the "inter-stencil redundancies" pass because disabling that pass makes everything else work fine in speculative mode.

**Comment by FabioLuporini** [comment]
> after a week of looking



> The biggest problem in this branch currently is that test_dse.py/test_tti_rewrite_speculative fails because there are some hidden assumptions about all loops running from 0 to dimension_size in the speculative rewrite passes of the DSE that I haven't been able to find after a week of looking. Can @FabioLuporini please look at making this test pass? More specifically, this problem is most probably localised within the "inter-stencil redundancies" pass because disabling that pass makes everything else work fine in speculative mode.

if you get stuck on something where I might help for more than a few hours, you should ask, not wait for 8x5=40 hours. Anyway, sure, I'll take a look. I'll do that once the branch has been rebased and we're sure that all other tests pass. Can you address the other comments, meanwhile?

**Comment by navjotk** [comment]
I did ask you when you were here. Twice. The first time you just explained to me how the pass works instead of helping me. The second time you said (and I'm paraphrasing) you didn't have the (mental) bandwidth to help me so I went back to trying to figure it out myself.

File: devito/profiling.py (hunk 12 above)
Comment (FabioLuporini) [reply devito/profiling.py]: reminder. This still needs to be addressed.

File: devito/operator.py (hunk 19 above)
Comment (FabioLuporini) [reply devito/operator.py]: any updates ?

**Comment by FabioLuporini** [comment]
I'm planning to take a look at the failing test case this afternoon

File: devito/operator.py (hunk 20 above)
Comment (navjotk) [reply devito/operator.py]: This will go in first

File: devito/operator.py (hunk 18 above)
Comment (navjotk) [reply devito/operator.py]: And that's exactly what the ArgumentEngine is slated to do

File: devito/operator.py (hunk 19 above)
Comment (navjotk) [reply devito/operator.py]: The ArgumentEngine will provide a runtime environment for the operator. I see only a cleaner interface because of its introduction - there is no overcomplication. The stencil radius is a property that is computed at compile (operator creation) time and used at runtime (apply). The ArgumentEngine is self-sufficient in what it needs to apply the operator and hence stores these properties on itself.

Comment (FabioLuporini) [reply devito/operator.py]: > What happens if you try using this?

? -- point is, if this suffices, we can just store this info and forget about all the rest

File: tests/test_autotuner.py [hunk 24]
```diff
@@ -102,7 +102,7 @@ def test_timesteps_per_at_run():
         op = Operator(stencil, dle=('blocking', {'blockalways': True}))
         op(infield=infield, outfield=outfield, autotune=True)
         out = [i for i in buffer.getvalue().split('\n') if 'AutoTuner:' in i]
-        expected = options['at_squeezer'] - to
```
Comment (FabioLuporini) [inline tests/test_autotuner.py]: why is this necessary?

File: devito/profiling.py (hunk 12 above)
Comment (ggorman) [reply devito/profiling.py]: Let me pick this apart. @FabioLuporini are you saying that the original code is wrong? Or is this functionally the same as the previous code but you have a different feeling for how it should be styled? If so, can we distinguish between advisory and required changes and style changes can potentially be put into a separate PR and applied consistently. Essentially (and particularly for large PR's) we don't want to have moving goalposts.

Comment (FabioLuporini) [reply devito/profiling.py]: I originally pointed out something about style (the ``dataspace`` calculation). Then I thought that even more important would be to verify that the "Performance Summary" is correct. For example, if we run ``acoustic`` on ``master`` and then on this branch, does Devito produce/display the same numbers (itershape, datashape, ..)? In the past, @mloubout and I had to do some serious investigation to find out why the performance numbers were off, so I just wanna make sure that we're still producing exactly the same reports as ``master`` (hence my "reminder" from a few days ago).

BTW, the style change was a one-line edit. I'm not such a great believer in "this low-priority thing will be improved in a subsequent PR", because it rarely happens in practice. So it's much better to do it straight away. This is the philosophy we've adopted since the Devito3.0 rewrite (Mathias can confirm)

File: devito/operator.py (hunk 20 above)
Comment (ggorman) [reply devito/operator.py]: @FabioLuporini - do you mean "drop the patch" or "drop the review comment"?

File: devito/operator.py (hunk 18 above)
Comment (ggorman) [reply devito/operator.py]: What's the conclusion from this discussion? Are there changes required to this PR or is the increasing complexity of .arguments being highlighted for some TLC (ie should an issue be articulated explaining the problem and we can assign the task to someone)?

File: devito/operator.py (hunk 20 above)
Comment (FabioLuporini) [reply devito/operator.py]: I was assuming #373 would go in first (it's ready, while this one is not); if #373 goes in first, we can drop the ``is_Fixed`` check above, because #373 removes ``FixedDimension``

**Comment by FabioLuporini** [comment]
After a thorough offline discussion, we're in this situation:

- The branch ``int_start_end_fabio`` has a commit that should fix some of the current Travis failures
- However, even with this patch, the current PR still breaks code generation in various places, in particular due to dropping offsets on the floor during ``_schedule_expressions`` (the routine responsible for constructing loops). I think I convinced @navjotk that this is wrong, but still we need a solution to the following problem:

Say we want an Operator to run for 20 iterations along the time dimension, starting at timestep 50 (doesn't have to be time, but just to make the discussion more concrete... as time is the driving use case of the PR). We could easily generate code such as:
```
for (t = t_start + 1 ; t < t_end - 1; t++) {
  ...
}
```
The offsets +-1 are necessary to avoid out-of-bounds array accesses along the time dimension.
So, what does it mean "to start executing at timestep 50" ? Does it mean ``t_start`` should be ``== 49``? Or should we actually start from ``t=51``? What convention should we follow?

The PR as is in its current state was trying to work around this issue by dropping the offsets in the generated code (perhaps I'm missing some details here, please @navjotk  clarify as necessary). But, as I said, we do agree this is in general wrong (we do need offsets, for a number of reasons).

So I guess we need a convention before a fix ( @ggorman @mlange05 @mloubout warmly invited to contribute).

**Comment by mlange05** [comment]
Ok, I generally think this ties in with the upcoming switch to properly using the concept of a "domain", as termed in YASK. This domain does not include the "boundary"/"padding" region that arises due to the stencil and effectively constitutes the part of the iteration that is written to. I think the user should address iteration indices with respect to the "domain", which might be different to the actual data indices in memory. To illustrate, some ASCII art:

```
 boundary             domain                     boundary
  | -- | ---------------------------------------- | -- |
   read                write                       read

Data indices:
  0    1               51     71                 101  102
User indices
       0               50     70                 100
```

I think the example code @FabioLuporini showed above would exactly achieve this, no?

**Comment by navjotk** [comment]
I would call the offline discussion inconclusive. 
I have since merged the branch with the DSE fixes and added a test to show that Iterations are still perfectly capable of producing offsets if you choose to use them for some bit of the code.

With code like: 
```
for (t = t_start + 1 ; t < t_end - 1; t++) {
  ...
}
```
we are lying about which timestep we are running since, being passed `t_start=50`, we would start with `t=51`. To make this more transparent to the user, this PR now generates code like:
```
for (t = t_start; t < t_end; t++) {
  ...
}
``` 
where `t_start` and `t_end` already carry the effect of the offset at run time. 

I do not agree that it is wrong to do this since logically we are still doing the same thing. 

Specifically, I don't see why certain DSE passes should break because of this, when it works fine with the lower level DSE optimisations. I continue to insist that this is a bug in the DSE being exposed as a side-effect of the PR. 

I am not in favour of dragging along this PR into a never-ending discussion.

@mlange05 I don't see your point

**Comment by FabioLuporini** [comment]
> and added a test to show that Iterations are still perfectly capable of producing offsets if you choose to use them for some bit of the code.

Of course they are: you haven't touched the ``Iteration`` implementation. When I say that you're dropping the offsets on the floor, what I mean is that now an ``Operator`` (through ``_schedule_expressions``), which is the only entity in Devito constructing ``Iteration``s (apart from some DLE passes), is "dropping them", but ignoring the information carried by a ``cluster``. This is the incriminating line:

https://github.com/opesci/devito/pull/345/files#diff-de8fe5c64c6d55d55d54343f37e2397cR321

in particular, we're not using ``j.ofs`` anymore (don't forget that ``is_Fixed`` is now totally irrelevant).

> Specifically, I don't see why certain DSE passes should break because of this, when it works fine with the other DSE optimisations. I continue to insist that this is a bug in the DSE being exposed as a side-effect of the PR.

We basically spent the whole of yesterday talking about this, about why we need offsets in Iterations. I thought I had convinced you, but I may be wrong. Different ``Iterations `` over the same dimension, say ``x``, may require slightly different iteration ranges. For example, in the same kernel, I may have:
```
for (x = 1 ; x < x_size -1 ; x++)
...
...
for (x = 2 ; x < x_size -2 ; x++)
```
It's logically still the `x` dimension (it is and it must be); but the loop bounds of the first loop lead to a slightly larger iteration range because of DSE-induced optimizations. 

So you cannot just replace both of them with a single ``x_start``; it won't work, because the first loop/iteration differs from the second one. You might use ``x_start`` and ``x_start + 1``; or you might write a completely new implementation (but why) where you have ``x_start_1`` and ``x_start_2`` (with the property that ``x_start_2 == x_start_1 + 1``; or whatever. 

So, again, this issue is *not* exposing any DSE bug. The DSE works as it's been conceived. We might argue about how bad or cool is it, but I don't see why this should be called a DSE bug.

On the other hand, the bug is in the current PR implementation (as I hope I have clarified)

> @mlange05 I don't see your point

I do. And given that and what I just wrote, I don't see why a solution in which we write loops such as:
```
for (x = 1 + x_start; x < x_end -1 ; x++)
...
...
for (x = 2 + x_start ; x < x_end -2 ; x++)
```
wouldn't just work (compare to the other snippet above; I added ``x_start`` here to both loops). ``x_start`` is user-provided at apply time (i.e., ``op.apply(x_start=50)``, as expected, I think, by the current API proposal). Michael basically said: ignore the offsets.

**Comment by mlange05** [comment]
> we are lying about which timestep we are running since, being passed t_start=50, we would start with t=51.

FWIW, I think this is only true for time due to the fact that we currently treat time as a boundary-less dimension (because of `save`, etc). In this case `j.ofs` should always be 0 for time and we might just force this behaviour. For other dimension types, eg. the space dimensions, "lying about" the iteration index is actually the aim here, although I admit that we haven't switched to the "domain" concept yet and are currently still treating everything as absolute indices (from a user perspective).

Regardless of this, the mechanism to consider offsets will be needed as soon as we do switch, and is already required due to the inter-stencil redundancy stuff (if I understand @FabioLuporini correctly).

>  (don't forget that is_Fixed is now totally irrelevant).

Is this fact affecting the current discussion now? If so, should #373 be held on account of this PR, or should we re-evaluate this PR in light of #373? @navjotk

**Comment by FabioLuporini** [comment]
> Regardless of this, the mechanism to consider offsets will be needed as soon as we do switch, and is already required due to the inter-stencil redundancy stuff (if I understand @FabioLuporini correctly).

So I wouldn't say "required", strictly speaking, because things as they are now on master are working as they're meant to be. Just for completeness of the discussion: ``Cluster``s, through ``Stencil``s, carry the offset information prior to loop construction; ``Cluster``s are originally created by the ``Operator``, based on the user-provided equations, and subsequently "modified" by the DSE (which may add or drop clusters). 

The offset/domain/... mechanism, through a ``Grid``, will be fundamental for YASK.

**Comment by mloubout** [comment]
Shouldn't this PR also fix/be related #69 ?
The one thing to remember is that, if the `start/end` are called multiple times in a row, the only thing that should matter is that you get the same thing as running the full time/space at once, if not there is a big problem.

>we are lying about which timestep we are running since, being passed t_start=50, we would start with t=51.

this will depend on the kernel/PDE/discretization, ignoring offset is risky down the road and I think whatever `tstart` or `xstart` and same for end should correspond to the ones computed so for acoustic:

```
for (t = t_start-1; t < t_end; t++) {
  u[t+1] = .... (first computed is t_start, last is t_end)
}
t_start_new = t_end+1
t_end_new=....
```

but with symbolic check to insure the value passed are the correct one and don't go out of bounds.

**Comment by mlange05** [comment]
> > we are lying about which timestep we are running since, being passed t_start=50, we would start with t=51.

> FWIW, I think this is only true for time due to the fact that we currently treat time as a boundary-less dimension 

@navjot Would it make things easier if we introduced explicit `TimeDimension` types now? I have them in a separate WIP branch atm, but they can be fast-tracked. With this you could simply do something like
```
iters = [Iteration([], j.dim, j.dim.limits, offsets=(0, 0) if j.dim.is_Time else j.ofs)
```

**Comment by FabioLuporini** [comment]
> @Navjot Would it make things easier if we introduced explicit TimeDimension types now? I have them in a separate WIP branch atm, but they can be fast-tracked. With this you could simply do something like

If we do 0-offsets in time, can we already guarantee that we generate something like:
```
u[t+k] = f(u[t+k-1], u[t+k-2], ..., u[t])
```
rather than, for example:
```
u[t] = f(u[t-1], u[t-2], ..., u[t-k])
```
ie, we don't end up with OOB accesses?

**Comment by mlange05** [comment]
OK, #382 has landed, so this branch needs a rebase/merge. @navjotk

File: devito/profiling.py [hunk 25]
```diff
@@ -137,17 +137,19 @@ def summary(self, dim_sizes, dtype):
...
             gpoints = iterspace/10**9
 
             # Compulsory traffic
-            datashape = [dim_sizes[dims[i].name] for i in itspace]
+            datashape = [(arguments[dims[i].end_name] - arguments[dims[i].start_name])
+                         for i in itspace]
             dataspace = reduce(operator.mul, datashape)
             traffic = profile.memory*dataspace*dtype().itemsize
-
+            traffic = float(profile.memory*dataspace*dtype().itemsize)
```
Comment (FabioLuporini) [inline devito/profiling.py]: redundant line, use float(...) in the line above

Comment (navjotk) [reply devito/profiling.py]: merge leftover. fixed.

**Comment by navjotk** [comment]
Can this be merged?

**Comment by FabioLuporini** [comment]
please go
